### PR TITLE
Implement `BODYSTRUCTURE`

### PR DESCRIPTION
**Pull request by avitex**
Still working on this for #42, not yet ready to merge.
Just open to comments at this point.

On a side note, @djc, would you like me to split out changes to core.rs in a separate PR?
Just a few changes for refactoring available if you like.

TODO:
- [x] Impl message
- [x] Impl multipart
- [x] More tests

### CODE REVIEW

**Comment by avitex** [comment]
@djc A couple additions to the above question:

- What's the reasoning behind putting `body.rs` outside the parser module?
- What do you think about the meda-types specified in the RFC to be in an enum, with a custom variant?
- What's the reasoning behind utf8 and non-utf8 strings? Does it make sense to make all utf8, or none at all?

I'm sure I'll think of some more soon. Also note body structure depends on envelope, which I've split up from the message attr.

Cheers!

**Review by avitex — COMMENTED** [review]
I'm not sure I like what cargo-fmt has done here, looks very messy

File: imap-proto/src/types.rs [hunk 1]
```diff
@@ -111,13 +111,101 @@ pub enum SectionPath {
     Part(Vec<u32>, Option<MessageSection>),
 }
 
+#[derive(Debug, Eq, PartialEq)]
+pub struct BodyParam<'a> {
```
Comment (avitex) [inline imap-proto/src/types.rs:115]: @djc Would you prefer body param as is here, or as a tuple?

Comment (djc) [reply imap-proto/src/types.rs:115]: I think a tuple is probably fine for this.

File: imap-proto/src/types.rs [hunk 2]
```diff
@@ -111,13 +111,101 @@ pub enum SectionPath {
...
+
+#[derive(Debug, Eq, PartialEq)]
+pub enum BodyExtension<'a> {
+    Num(u32),
+    Str(Option<&'a str>),
+    List(Vec<BodyExtension<'a>>),
+}
+
+#[derive(Debug, Eq, PartialEq)]
+pub struct BodyStructureText<'a> {
```
Comment (djc) [inline imap-proto/src/types.rs:134]: There is a lot of duplication between the different `BodyStructure` types. Can you think about a way to reduce the duplication here? Not yet sure what that would look like, but I wouldn't want to merge it as is.

File: imap-proto/src/types.rs [hunk 3]
```diff
@@ -111,13 +111,101 @@ pub enum SectionPath {
...
+}
+
 #[derive(Debug, Eq, PartialEq)]
 pub enum AttributeValue<'a> {
     BodySection {
         section: Option<SectionPath>,
         index: Option<u32>,
         data: Option<&'a [u8]>,
     },
+    BodyStructure(Box<BodyStructure<'a>>),
```
Comment (djc) [inline imap-proto/src/types.rs:208]: Why do we have the `Box` here?

File: imap-proto/src/core.rs [hunk 4]
```diff
@@ -87,6 +87,10 @@ named!(pub nstring<Option<&[u8]>>, alt!(
     map!(string, |s| Some(s))
 ));
 
+named!(pub string_utf8<&str>, map!(string, |s| str::from_utf8(s).unwrap()));
```
Comment (djc) [inline imap-proto/src/core.rs]: It'd be nice if you introduce these helpers to convert existing `str::from_utf8()` users that would benefit from these. Ideally in that case the whole change would go in a separate commit.

File: imap-proto/src/types.rs [hunk 5]
```diff
@@ -111,13 +111,69 @@ pub enum SectionPath {
...
+        octets: u32,
+        md5: Option<&'a str>,
+        disposition: Option<BodyDisposition<'a>>,
+        lang: Option<Vec<&'a str>>,
+        loc: Option<&'a str>,
+        lines: u32,
+        extensions: Option<BodyExtension<'a>>
+    },
+    // TODO
+    Message,
```
Comment (djc) [inline imap-proto/src/types.rs]: Please fold the commits that add these together into this one.

File: imap-proto/src/core.rs [hunk 6]
```diff
@@ -59,6 +59,12 @@ pub fn quoted_data(i: &[u8]) -> IResult<&[u8], &[u8]> {
     Ok((&i[len..], &i[..len]))
 }
 
+/// space = a single space character
+named!(pub space, tag_s!(" "));
```
Comment (djc) [inline imap-proto/src/core.rs:63]: What's the reason to extract this? It doesn't seem to be a separate syntax item in the RFC, and in this case it feels like the abstraction doesn't add much value beyond the 6-character reduction.

File: imap-proto/src/core.rs [hunk 7]
```diff
@@ -59,6 +59,12 @@ pub fn quoted_data(i: &[u8]) -> IResult<&[u8], &[u8]> {
...
+
+// nil = "NIL"
+named!(pub nil, tag_s!("NIL"));
```
Comment (djc) [inline imap-proto/src/core.rs:66]: I'd be hesitant about this, too, but since it's in the syntax it might be helpful.

File: imap-proto/src/core.rs [hunk 8]
```diff
@@ -81,29 +87,28 @@ named!(pub literal<&[u8]>, do_parse!(
 /// string = quoted / literal
 named!(pub string<&[u8]>, alt!(quoted | literal));
 
+/// string bytes as as utf8
+named!(pub string_utf8<&str>, map_res!(string, str::from_utf8));
+
 /// nstring = string / nil
 named!(pub nstring<Option<&[u8]>>, alt!(
-    map!(tag_s!("NIL"), |_| None) |
+    map!(nil, |_| None) |
```
Comment (djc) [inline imap-proto/src/core.rs:95]: But if you're going to convert existing uses, please make it into a separate commit.

File: imap-proto/src/core.rs [hunk 9]
```diff
@@ -81,29 +87,28 @@ named!(pub literal<&[u8]>, do_parse!(
...
     map!(string, |s| Some(s))
 ));
 
-named!(pub string_utf8<&str>, map!(string, |s| str::from_utf8(s).unwrap()));
-
-named!(pub nstring_utf8<Option<&str>>, map!(nstring, |val| val.map(|s| str::from_utf8(s).unwrap())));
+/// nstring bytes as utf8
```
Comment (djc) [inline imap-proto/src/core.rs:99]: Please don't mix code movement or reformatting in the same commit as functional changes, this makes it much harder for me to review.

**Review by djc — COMMENTED** [review]
Thanks for all your work!

For a larger, more complex PR like this I'm going to be a bit nitpicky about how it's represented so I can do a proper review -- I hope that's okay. What I'd like to see here:

* Meaningful, coherent commits that make a small single change each, each of which passes the tests (including rustfmt)
* No flip-flopping: don't make small/non-functional changes in later commits to code from earlier commits

I recommend using interactive rebase to split and/or squash your existing commits, but given the sequence you got now it might be a bit tricky. In these cases sometimes I just start from scratch and redo the entire sequence (with cherry-picking of commits and some copy/paste work where it makes sense). Let me know if you need some tips on how to do that.

Otherwise it mostly looks good, although I really want to figure out a neater way of abstracting between the different body structure types.

**Comment by avitex** [comment]
@djc Could you please have a look at my questions in my [above comment](https://github.com/djc/tokio-imap/pull/44#issuecomment-485221137).
I'm happy to pull out the changes from the core in a separate PR, made amends to this one and squash.

**Comment by djc** [comment]
Enum for media types could make sense, what section of the RFC are you referring to exactly?

As for UTF-8 vs bytes, I think I was going for string types where it seems likely or is specified that no non-UTF-8 content will ever happen (I think some fields are specified as ASCII) and bytes for the rest (especially for large chunks that include email bodies).

**Comment by avitex** [comment]
> Enum for media types could make sense, what section of the RFC are you referring to exactly?

There are a few different specific media-types specified in the [formal syntax](https://tools.ietf.org/html/rfc3501#section-9) - see `media-basic` for example

> As for UTF-8 vs bytes, I think I was going for string types where it seems likely or is specified that no non-UTF-8 content will ever happen (I think some fields are specified as ASCII) and bytes for the rest (especially for large chunks that include email bodies).

Seems fair enough; I'll have to have to look into more detail on the RFC for that.

**Comment by djc** [comment]
IIRC there are crates around mime types already, maybe we should import their types rather than defining our own?

**Comment by avitex** [comment]
Want to look at that as a separate PR?

**Comment by djc** [comment]
Yeah, that would probably make sense.

**Comment by avitex** [comment]
Closing in favour of #48. I have a dependency on this branch.

**Comment by lberezy** [comment]
Closes #42
